site/src/components/Resources/AppLink/AppLink.tsx

```diff
@@ -98,11 +103,46 @@ export const AppLink: FC<AppLinkProps> = ({ app, workspace, agent }) => {
             canClick
               ? (event) => {
                   event.preventDefault();
-                  window.open(
-                    href,
-                    Language.appTitle(appDisplayName, generateRandomString(12)),
-                    "width=900,height=600",
-                  );
+                  // This is an external URI like "vscode://", so
+                  // it needs to be opened with the browser protocol handler.
+                  if (app.external && !app.url.startsWith("http")) {
+                    // If the protocol is external the browser does not
+                    // redirect the user from the page.
+
+                    // This is a magic undocumented string that is replaced
+                    // with a brand-new session token from the backend.
+                    // This only exists for external URLs, and should only
+                    // be used internally, and is highly subject to break.
+                    const magicTokenString = "$SESSION_TOKEN";
+                    const hasMagicToken = href.indexOf(magicTokenString);
+                    if (hasMagicToken !== -1) {
+                      setFetchingSessionToken(true);
+                      getApiKey()
+                        .then((key) => {
+                          const url = href.replaceAll(
+                            magicTokenString,
+                            key.key,
+                          );
+                          window.location.href = url;
+                          setFetchingSessionToken(false);
+                        })
+                        .catch((ex) => {
+                          console.error(ex);
+                          setFetchingSessionToken(false);
+                        });
```

Comment:
you could just make this event handler `async` and then `await` these

Reply:
I suppose that's the same as console.error really, will change